### PR TITLE
Rambotzhong/spring client auth

### PR DESCRIPTION
**Pull request by rambotzhong**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
The precondition is to put your code inside this repository and upgrade the version in the pull request but do not modify `src/index.json`.

### CODE REVIEW

**Review by bowen5 — APPROVED** [review]

**Comment by yonzhan** [comment]
spring

File: src/spring/azext_spring/_app_factory.py [hunk 1]
```diff
@@ -130,15 +132,19 @@ def _load_vnet_addons(self, public_for_vnet=None, **_):
...
+                (session_affinity is not None) or (session_max_age is not None) or (backend_protocol is not None) or \
+                (client_auth_certs is not None):
             return models_20220901preview.IngressSettings(
                 read_timeout_in_seconds=ingress_read_timeout,
                 send_timeout_in_seconds=ingress_send_timeout,
                 session_affinity=session_affinity,
                 session_cookie_max_age=session_max_age,
-                backend_protocol=backend_protocol
+                backend_protocol=backend_protocol,
+                client_auth=IngressSettingsClientAuth(
```
Comment (yuwzho) [inline src/spring/azext_spring/_app_factory.py:145]:
```suggestion
                client_auth=models_20220901preview.IngressSettingsClientAuth(
```

File: src/spring/azext_spring/_app_factory.py [hunk 2]
```diff
@@ -11,6 +11,8 @@
 from .vendored_sdks.appplatform.v2022_09_01_preview import models as models_20220901preview
 from azure.cli.core.util import get_file_json
 
+from .vendored_sdks.appplatform.v2022_09_01_preview.models import IngressSettingsClientAuth
```
Comment (yuwzho) [inline src/spring/azext_spring/_app_factory.py:14]:
```suggestion

```

File: src/spring/azext_spring/_app_factory.py [hunk 3]
```diff
@@ -130,15 +132,19 @@ def _load_vnet_addons(self, public_for_vnet=None, **_):
...
             return models_20220901preview.IngressSettings(
                 read_timeout_in_seconds=ingress_read_timeout,
                 send_timeout_in_seconds=ingress_send_timeout,
                 session_affinity=session_affinity,
                 session_cookie_max_age=session_max_age,
-                backend_protocol=backend_protocol
+                backend_protocol=backend_protocol,
+                client_auth=IngressSettingsClientAuth(
+                    certificates=client_auth_certs
+                ) if client_auth_certs is not None and len(client_auth_certs) > 0 else None
```
Comment (yuwzho) [inline src/spring/azext_spring/_app_factory.py:147]:
```suggestion
                ) if client_auth_certs else None
```

File: src/spring/azext_spring/_params.py [hunk 4]
```diff
@@ -402,6 +403,10 @@ def prepare_logs_argument(c):
             c.argument('backend_protocol',
                        arg_type=get_enum_type(BackendProtocol),
                        help='Ingress backend protocol of app.')
+            c.argument('client_auth_certs',
+                       type=str,
+                       validator=validate_ingress_client_auth_certificates,
+                       help="A space-separated string containing resource ids of certificates for client authentication. e.g: --client_auth_certs='id0 id1'. Use '' to clear existing certificates.")
```
Comment (yuwzho) [inline src/spring/azext_spring/_params.py:409]:
```suggestion
            c.argument('client_auth_certs',
                       nargs='*',
                       validator=validate_ingress_client_auth_certificates,
                       help="A space-separated string containing resource ids of certificates for client authentication. e.g: --client_auth_certs='id0 id1'. Use '' to clear existing certificates.")
```

Comment (yuwzho) [reply src/spring/azext_spring/_params.py:409]: This can help you to split the arguments

**Comment by yuwzho** [comment]
You need to update setup.py to increase the version

File: src/spring/azext_spring/_app_factory.py (hunk 3 above)
Comment (rambotzhong) [reply src/spring/azext_spring/_app_factory.py]: have updated to the judge the if the client_auth-certs is not None, if the is empty list, we will also set it and request the backend service

**Review by bowen5 — APPROVED** [review]

**Review by bowen5 — APPROVED** [review]

**Review by bowen5 — APPROVED** [review]

**Comment by zhoxing-ms** [comment]
Could you please add some tests for the new parameter?

**Review by yuwzho — APPROVED** [review]

**Comment by rambotzhong** [comment]
> Could you please add some tests for the new parameter?

added, thanks for reminding

**Comment by rambotzhong** [comment]
hi @zhoxing-ms  @yanzhudd  would you please help to review?

File: src/spring/azext_spring/_params.py (hunk 4 above)
Comment (zhoxing-ms) [inline src/spring/azext_spring/_params.py:409]: `type=str` is the default setting, so it does not need to be explicitly specified
```suggestion
            c.argument('client_auth_certs',
                       validator=validate_ingress_client_auth_certificates,
                       help="A space-separated string containing resource ids of certificates for client authentication. e.g: --client_auth_certs='id0 id1'. Use '' to clear existing certificates.")
```

**Comment by zhoxing-ms** [comment]
Could you add some scenario test for the new parameter?

File: src/spring/azext_spring/_params.py (hunk 4 above)
Comment (rambotzhong) [reply src/spring/azext_spring/_params.py]: thanks, updated

File: src/spring/azext_spring/tests/latest/test_asa_app_scenario.py [hunk 5]
```diff
@@ -288,4 +288,26 @@ def test_vnet_public_endpoint(self):
 
         self.cmd('spring app update -n {app} -g {rg} -s {serviceName} --assign-public-endpoint true', checks=[
             self.check('properties.vnetAddons.publicEndpoint', True)
-        ])
\ No newline at end of file
+        ])
+
+
+@record_only()
```
Comment (zhoxing-ms) [inline src/spring/azext_spring/tests/latest/test_asa_app_scenario.py:294]: May I ask why this test marked as `record_only`?

Comment (rambotzhong) [reply src/spring/azext_spring/tests/latest/test_asa_app_scenario.py:294]: the scenario depends on the creation of Azure Spring Apps, which is time-consuming and might be failed caused by backend or lack of resources. So I marked it as record_only to using the recording test.

Comment (zhoxing-ms) [reply src/spring/azext_spring/tests/latest/test_asa_app_scenario.py:294]: OK, got it~

**Review by zhoxing-ms — APPROVED** [review]

**Pull request merged**